Sprint 6 solution

shinket7 opened the pull request.

shinket7 committed:
Create `remove()` in HistoryManager interface and its implementation.

shinket7 committed:
change signature `getHistory()` of `HistoryManager` interface.

Now it returns `List` instead of `ArrayList`.

shinket7 committed:
Add test for `remove()` of `InMemoryHistoryManager`.

Add new test.
Change types of variables of expected lists of history in `InMemoryHistoryTest` tests.

shinket7 committed:
Add test for `hashCode()` of Task model.

shinket7 committed:
Create `SinglyLinkedTaskList`. Create `TaskNode` for it.

shinket7 committed:
Change history to usage of `SinglyLinkedTaskList` and historyMap.

Update `SinglyLinkedTaskList`: make it fully static.
Remove history length restriction. Make history unique.

shinket7 committed:
Unmake `SinglyLinkedTaskList` static back. Fix other bugs in history.

Unmake `SinglyLinkedTaskList` static back.
Remove history length restriction tests.
Change `linkLast()` argument from `Task` to `TaskNode`.
Move `removeNode()` into `SinglyLinkedTaskList`, add `size` reduction.

shinket7 committed:
Fix history `add()` and its `removeNode()`. Update demo in `Main`.

shinket7 committed:
Move `TaskNode` class to its own file.

shinket7 committed:
Move `Demo` class to its own file.

avfyodorov commented on src/tracker/controllers/TaskNode.java:
```diff
@@ -0,0 +1,36 @@
+
+import tracker.model.Task;
+
+public class TaskNode {
```
Этот класс можно было бы сделать внутренним классом менеджера истории, потому что он всё равно больше нигде не нужен. 

На усмотрение.

shinket7 replied:
Перенёс во внутрь класса `InMemoryHistoryManager`, чтобы было в одном месте.

avfyodorov commented on src/tracker/controllers/InMemoryHistoryManager.java:
```diff
@@ -3,24 +3,94 @@
+
+        private TaskNode first;
+        private TaskNode last;
+        private int size;
```
`private int size;`

Это поле точно не нужно, для того чтобы узнать размер истории можно воспользоваться методом 

historyMap.size()

shinket7 replied:
Заменил на использование `size()` от мапы в `historyMap`. Убрал поле `size` из `SinglyLinkedTaskList`.

avfyodorov commented on src/tracker/controllers/InMemoryTaskManager.java:
```diff
@@ -7,6 +7,7 @@
+import java.util.List;
 
 public class InMemoryTaskManager implements TaskManager {
 
```
По поводу удаления. Пожалуйста, обратите внимание, что во всех методах, отвечающих за удаление задач, подзадач и эпиков также необходимо удалять их и из истории.

shinket7 replied:
Добавил удаление из истории при удалении задач.
Добавил тесты на эти кейсы.

avfyodorov reviewed (commented):
Добрый день, Виталий! 

В целом хорошо, правда, есть ряд уточнений.

shinket7 committed:
Make `TaskNode` class and inner class of `InMemoryHistoryManager`.

shinket7 committed:
Remove `size` field from `SinglyLinkedTaskList`.

shinket7 committed:
Make deletion of tasks delete them from history.

shinket7 committed:
Add tests for history deletion when tasks are deleted.

Add tests for clear and delete methods of `InMemoryTaskManager` that deleted tasks are deleted from history too.
Fix `removeNode()` issue when remove the last node from history.

shinket7 committed:
Fix small typo in test message in `TaskTest`.

shinket7 committed:
Fix typo in test message in `InMemoryTaskManagerTest`.

avfyodorov reviewed (commented):
Добрый день, Виталий!

Замечаний нет.
Работа принята.
